// @ldo/type-traverser/dist/transformer/transformerSubTraversers/TransformerParentSubTraverser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transformerParentSubTraverser = void 0;
const TransformerInterfaceSubTraverser_1 = require("./TransformerInterfaceSubTraverser");
const TransformerPrimitiveSubTraverser_1 = require("./TransformerPrimitiveSubTraverser");
const TransformerUnionSubTraverser_1 = require("./TransformerUnionSubTraverser");
const timeout_1 = require("./util/timeout");
const subTraversers = {
    interface: TransformerInterfaceSubTraverser_1.transformerInterfaceSubTraverser,
    union: TransformerUnionSubTraverser_1.transformerUnionSubTraverser,
    primitive: TransformerPrimitiveSubTraverser_1.transformerPrimitiveSubTraverser,
};
async function transformerParentSubTraverser(item, itemTypeName, globals) {
    const { traverserDefinition, executingPromises } = globals;
    if (executingPromises.has(item, itemTypeName)) {
        return executingPromises.get(item, itemTypeName)?.promise;
    }
    const subTraverser = subTraversers[traverserDefinition[itemTypeName].kind];
    const executingPromise = {
        promise: (async () => {
            await (0, timeout_1.timeout)(0);
            return subTraverser(item, itemTypeName, globals);
        })(),
        isResolved: false,
    };
    executingPromises.set(item, itemTypeName, executingPromise);
    const toReturn = await executingPromise.promise;
    executingPromise.isResolved = true;
    return toReturn;
}
exports.transformerParentSubTraverser = transformerParentSubTraverser;
