// @ldo/type-traverser/dist/instanceGraph/nodes/InstanceNode.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InstanceNode = void 0;
class InstanceNode {
    constructor(graph, instance, typeName) {
        this.parents = {};
        this.graph = graph;
        this.instance = instance;
        this.typeName = typeName;
    }
    getParentKey(identifiers) {
        return identifiers.join("|");
    }
    _setParent(identifiers, parentNode) {
        const parentKey = this.getParentKey(identifiers);
        if (!this.parents[parentKey])
            this.parents[parentKey] = new Set();
        this.parents[parentKey].add(parentNode);
    }
    parent(...identifiers) {
        return Array.from(this.parents[this.getParentKey(identifiers)] ?? []);
    }
    allParents() {
        return Object.values(this.parents)
            .map((parentSet) => Array.from(parentSet))
            .flat();
    }
    get traverserDefinition() {
        return this.graph.traverserDefinitions[this.typeName];
    }
}
exports.InstanceNode = InstanceNode;
