// @ldo/rdf-utils/dist/nodeSerialization.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.stringToQuadMatch = exports.stringToQuad = exports.stringToNode = exports.stringToGraphNode = exports.stringToObjectNode = exports.stringToPredicateNode = exports.stringToSubjectNode = exports.quadMatchToString = exports.quadToString = exports.nodeToString = exports.grpahNodeToString = exports.objectNodeToString = exports.predicateNodeToString = exports.subjectNodeToString = exports.AnyTermTypes = exports.GraphTermTypes = exports.ObjectTermTypes = exports.PredicateTermTypes = exports.SubjectTermTypes = void 0;
const rdf_string_1 = require("rdf-string");
exports.SubjectTermTypes = new Set([
    "NamedNode",
    "BlankNode",
]);
exports.PredicateTermTypes = new Set([
    "NamedNode",
]);
exports.ObjectTermTypes = new Set([
    "NamedNode",
    "BlankNode",
    "Literal",
]);
exports.GraphTermTypes = new Set([
    "NamedNode",
    "BlankNode",
    "DefaultGraph",
]);
exports.AnyTermTypes = new Set([
    "NamedNode",
    "BlankNode",
    "DefaultGraph",
    "Literal",
]);
function subjectNodeToString(node) {
    return nodeToString(node);
}
exports.subjectNodeToString = subjectNodeToString;
function predicateNodeToString(node) {
    return nodeToString(node);
}
exports.predicateNodeToString = predicateNodeToString;
function objectNodeToString(node) {
    return nodeToString(node);
}
exports.objectNodeToString = objectNodeToString;
function grpahNodeToString(node) {
    return nodeToString(node);
}
exports.grpahNodeToString = grpahNodeToString;
function nodeToString(node) {
    return (0, rdf_string_1.termToString)(node);
}
exports.nodeToString = nodeToString;
function quadToString(quad) {
    return JSON.stringify((0, rdf_string_1.quadToStringQuad)(quad));
}
exports.quadToString = quadToString;
function quadMatchToString(quadMatch) {
    return JSON.stringify({
        subject: quadMatch[0] ? subjectNodeToString(quadMatch[0]) : undefined,
        predicate: quadMatch[1] ? predicateNodeToString(quadMatch[1]) : undefined,
        object: quadMatch[2] ? objectNodeToString(quadMatch[2]) : undefined,
        graph: quadMatch[3] ? grpahNodeToString(quadMatch[3]) : undefined,
    });
}
exports.quadMatchToString = quadMatchToString;
function stringToSubjectNode(input) {
    return stringToNode(input, exports.SubjectTermTypes);
}
exports.stringToSubjectNode = stringToSubjectNode;
function stringToPredicateNode(input) {
    return stringToNode(input, exports.PredicateTermTypes);
}
exports.stringToPredicateNode = stringToPredicateNode;
function stringToObjectNode(input) {
    return stringToNode(input, exports.ObjectTermTypes);
}
exports.stringToObjectNode = stringToObjectNode;
function stringToGraphNode(input) {
    return stringToNode(input, exports.GraphTermTypes);
}
exports.stringToGraphNode = stringToGraphNode;
function stringToNode(input, expectTermType) {
    const node = (0, rdf_string_1.stringToTerm)(input);
    if (expectTermType && !expectTermType.has(node.termType)) {
        throw new Error(`Expected term to be one of term type: [${Array.from(expectTermType).reduce((agg, termType) => `${agg}${termType}, `, "")}], but got ${node.termType}.`);
    }
    return node;
}
exports.stringToNode = stringToNode;
function stringToQuad(input) {
    try {
        return (0, rdf_string_1.stringQuadToQuad)(JSON.parse(input));
    }
    catch (err) {
        throw new Error("Invalid Quad String");
    }
}
exports.stringToQuad = stringToQuad;
function stringToQuadMatch(input) {
    try {
        const jsonRep = JSON.parse(input);
        return [
            jsonRep.subject != undefined
                ? stringToSubjectNode(jsonRep.subject)
                : undefined,
            jsonRep.predicate != undefined
                ? stringToPredicateNode(jsonRep.predicate)
                : undefined,
            jsonRep.object != undefined
                ? stringToObjectNode(jsonRep.object)
                : undefined,
            jsonRep.graph != undefined ? stringToGraphNode(jsonRep.graph) : undefined,
        ];
    }
    catch (err) {
        throw new Error("Invalid Quad Match String");
    }
}
exports.stringToQuadMatch = stringToQuadMatch;
